feat: create social authentication guide (#4976)
cypress-io/cypress-documentation: docs/guides/end-to-end-testing/testing-strategies/social-authentication.mdx

---
title: Social Authentication
slug: /guides/end-to-end-testing/social-authentication
---

:::info

## <Icon name="graduation-cap"></Icon> What you'll learn

- Log into your application via Google, Facebook, or Microsoft through the UI
  with [`cy.origin()`](/api/commands/origin) using Auth0 social connections. The
  same patterns will work with Github, LinkedIn, and others.
- Understanding

>   [challenges](http://localhost:3000/guides/references/best-practices#When-logging-in)

Reviewer: should this be a localhost link?

  with social connections

:::

:::caution

<strong class="alert-header">
  <Icon name="exclamation-triangle"></Icon> Not recommended in CI
</strong>

Cypress does **not** recommend testing social connection authentication as a
primary means of application authentication testing. This is due to the
[challenges](/guides/references/best-practices#Challenges-with-Social-Connection-Authentication)
mentioned in our [Best Practices Guide](/guides/references/best-practice).

If testing user based authentication flows via social connections is a **must**
for your application, we recommend testing this in
[open mode](https://docs.cypress.io/guides/guides/command-line#cypress-open) (or
[run mode](https://docs.cypress.io/guides/guides/command-line#cypress-run)
locally and
[headed](https://docs.cypress.io/guides/guides/command-line#cypress-run-headed))
to avoid A/B tests or alternate login flows. Relying on social authentication in
CI will likely result in bot detection and in some cases, account suspension due
to violating the provider's Terms of Service.

:::

## Application Setup

For this guide, we are going to set up an application that uses Auth0 for

> authentication. Our Auth0 Tenant will have social connections set up for Google,

Reviewer: ```suggestion
authentication. Our Auth0 tenant will have social connections set up for Google,
```

Author: addressed in ac6af55da

> Facebook, and Microsoft which with authenticate users on our behalf to our

Reviewer: ```suggestion
Facebook, and Microsoft which will authenticate users on our behalf to our
```

Author: addressed in ac6af55da

application.

> ### Setting up the Tenant

Reviewer: ```suggestion
### Setting up the Auth0 tenant
```

Author: addressed in https://github.com/cypress-io/cypress-documentation/commit/ac6af55da2402a2a483d092323c97b44e6086712

- To get started, first create your free [Auth0](https://auth0.com/signup)
  account and select the `"I am a new Auth0 User"` flow. This will have 4 steps:
  - Step 1: When configuring your Sample App, please select the
    `Single-Page App` for you platform and`React` for the technology.
  - Step 2: For `Social Connections`, please select Facebook, Microsoft, and
    Google
  - Step 3: Click try login and verify the connections are set up properly
  - Step 4: You can either download the sample on the `Download Sample App`
    button or clone it
    [here](https://github.com/auth0-samples/auth0-react-samples/tree/d5848ed710db85d6ef060731b296603c0eca744b/Sample-01).
    We will need this application in the next step. Select

>     `"I am Ready to Use Auth0"`. Your social connections on your Auth0 Tenant

Reviewer: ```suggestion
    `"I am Ready to Use Auth0"`. Your social connections on your Auth0 tenant
```

Author: addressed in ac6af55da

    should now be configured correctly

> ### Setting up the Application

Reviewer: ```suggestion
### Setting up the application
```

Author: addressed in ac6af55da

    for `Application Type`,
  - Inside `Application URIs`, please add `http://localhost:3000` to
    `Allowed Callback URLs`, `Allowed Web Origins` and `Allowed Logout URLs`
- Create the `src/auth_config.json` file from the
  [`src/auth_config.json.example`](https://github.com/auth0-samples/auth0-react-samples/blob/d5848ed710db85d6ef060731b296603c0eca744b/Sample-01/src/auth_config.json.example)
  file as described in the
  [sample app](https://github.com/auth0-samples/auth0-react-samples/tree/d5848ed710db85d6ef060731b296603c0eca744b/Sample-01#create-an-api).
- Once created, paste the `domain` and `client ID` into their respective fields
  inside `src/auth_config.json`. The `audience` key can be removed. The
  login/callback URIs will default to `http://localhost:3000` which is the url
  of our test application.
- [install](https://github.com/auth0-samples/auth0-react-samples/tree/d5848ed710db85d6ef060731b296603c0eca744b/Sample-01#project-setup)
  the application dependencies and
  [start the dev server](https://github.com/auth0-samples/auth0-react-samples/tree/d5848ed710db85d6ef060731b296603c0eca744b/Sample-01#compile-and-hot-reload-for-development)

> - Verify the application is stood up correctly by visiting

Reviewer: ```suggestion
- Verify the application is running correctly by visiting
```

Author: addressed in ac6af55da

  `http://localhost:3000` and clicking the blue `Log in` button in the top right
  corner. Select one of our social connections and log in and make sure to
  complete the `Authorize App` prompt. This only needs to be done once. This
  should take you back to the sample app authenticated as the test user. When
  finished, click the `Log Out` dropdown button. Repeat for Google, Microsoft,
  and Facebook or the accounts you would like to test.

Here is what this process should look like for Facebook setup:

<DocsVideo src="https://vimeo.com/789093604"></DocsVideo>

## Testing with Cypress

> Now that our application is set up and ready to go, we are ready to write our

Reviewer: Does this section need more attention drawn to it?  Iconography or subheading perhaps?

Author: I am not sure. Usually with our other guides we tend to go with the H2 here, but I feel like in the others as well that the sections that matter somewhat get swallowed

      cy.contains('Next').click().wait(4000)
      cy.get('[type="password"]').type(password, {
        log: false,
      })
      cy.contains('Next').click().wait(4000)
    }
  )

  cy.get('h6.dropdown-header').should('contain', name)
}
```

And lastly, our Microsoft Login Function

```js
// cypress/support/commands.ts
import { domain as Auth0Domain } from '../../src/auth_config.json'

function logIntoMicrosoft(username: string, password: string, name: string) {
  cy.visit('http://localhost:3000')
  cy.get('#qsLoginBtn').click()

  cy.origin(Auth0Domain, () => {
    cy.scrollTo('bottom')
    cy.get('form[data-provider="windowslive"]').submit()
  })

  cy.origin(
    'login.live.com',
    {
      args: {
        username,
        password,
      },
    },
    ({ username, password }) => {
      cy.get('input[type="email"]').type(username)
      cy.get('input[type="submit"]').click()
      cy.get('input[type="password"]').type(password, {
        log: false,
      })
      cy.get('input[type="submit"]').click()
      cy.get('#idBtn_Back').click()
    }
  )

  cy.get('h6.dropdown-header').should('contain', name)
}
```

> We can wire together these functions in a single command, like so:

Reviewer: ```suggestion
We can wire these functions together in a single command, like so:
```

Author: addressed in ac6af55da

```js
// cypress/support/commands.ts

Cypress.Commands.add(
  'loginToAuth0ViaSocial',
  (SOCIAL_PROVIDER: 'microsoft' | 'google' | 'facebook') => {
    const log = Cypress.log({
      displayName: 'Social LOGIN',
      message: [`🔐 Authenticating | ${SOCIAL_PROVIDER}`],
      // @ts-ignore
      autoEnd: false,
    })
    log.snapshot('before')

    switch (SOCIAL_PROVIDER) {
      case 'microsoft':
        logIntoMicrosoft(
          Cypress.env('MICROSOFT_USERNAME'),
          Cypress.env('MICROSOFT_PASSWORD'),
          Cypress.env('MICROSOFT_NAME')
        )
        break
      case 'google':
        logIntoGoogle(
          Cypress.env('GOOGLE_USERNAME'),
          Cypress.env('GOOGLE_PASSWORD'),
          Cypress.env('GOOGLE_NAME')
        )
        break
      case 'facebook':
        logIntoFacebook(
          Cypress.env('FACEBOOK_USERNAME'),
          Cypress.env('FACEBOOK_PASSWORD'),
          Cypress.env('FACEBOOK_NAME')
        )
        break
      default:
        throw new Error('no social provider configured!')
    }

    log.snapshot('after')
    log.end()
  }
)
```

Now, we can use our `loginToAuth0ViaSocial` command in the test. Below is our

> test to login as a user via Auth0 and run a basic sanity check.

Reviewer: ```suggestion
test to login as a user via Auth0 and run a basic check.
```

Author: addressed in ac6af55da
